pints/_mcmc/__init__.py

```diff
@@ -293,13 +293,60 @@ class MCMCController(object):
     method : class
         The class of :class:`MCMCSampler` to use. If no method is specified,
         :class:`HaarioBardenetACMC` is used.
+    transform : pints.Transform
+        A :class:`pints.Transform` that transform the model parameter space to
+        search space. If no transform is specified, search will be performed
+        in the model space.
     """
 
-    def __init__(self, log_pdf, chains, x0, sigma0=None, method=None):
+    def __init__(
+            self, log_pdf, chains, x0, sigma0=None, method=None,
+            transform=None):
 
-        # Store function
+        # Check function
         if not isinstance(log_pdf, pints.LogPDF):
             raise ValueError('Given function must extend pints.LogPDF')
+
+        # Transform everything
+        # From this point onward the MCMC sampler will see only the
+        # transformed search space and will know nothing about the model
+        # parameter space.
+        if transform:
+            log_pdf = transform.apply_log_pdf(log_pdf)
+            x0 = [transform.to_search(x) for x in x0]
+            #
+            # Following Eq. 5 in
+            # http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.47.9023
+            #
+            # For transformation x = g(p) and Jacobian J(x) = d/dx g^{-1}(x),
+            # then the covariance matrices C(.) for x and p follow
+            #
+            # C(x) = (d/dp g(E(p))) C(p) (d/dp g(E(p)))^T
+            #        = J^{-1}(E(x)) C(p) J^{-1}(E(x))^T
+            #
+            # Using the property that J^{-1} = dg/dp, from the inverse function
+            # theorem, i.e. the matrix inverse of the Jacobian matrix of an
+            # invertible function is the Jacobian matrix of the inverse
+            # function.
+            #
+            if sigma0 is not None:
+                # Transform sigma0 if provided
```

Comment:
Around line 370 it says

>         # Don't check initial standard deviation: done by samplers!

That will need to go, if we _do_ check it here!
In that case we should separate the checking of sigma0 from the transforming

Reply:
And probably move x0 checking up here too then (or move transform checking down)

Reply:
Hmm yeah, but not quite doing the full check at the same time, it just need to handle those that's provided. If it's not provided, we can leave this to be handled later.

Reply:
If we're going to do it in the controller instead of the optimiser, I think it'd be less confusing for everybody if we did it all the time. Otherwise we're creating room for weird bugs and edge cases :-)

Reply:
I think I tried doing that, but that was a bit annoying for multi- and single-chain methods... but I guess I can give it another go and move all the checks etc. from the optimiser/sampler to the controller.

Reply:
In fact, what was the reason of having that in the optimiser/sampler instead of the controller in the first place?!

Reply:
@MichaelClerx It seems to be a bit annoying to move everything down to the controller, because we have `MultiChainMCMC` and `SingleChainMCMC` that do slightly different things... And wouldn't that lose the checks if someone decided to use the ask and tell with `MCMCSampler` -- as well as the handling of sigma0 if not provided!

Reply:
We will keep those checks in the samplers as they are for now, before we come up a better way of doing this such as moving them into a utility method...